server/src/server.rs

```diff
@@ -618,56 +819,83 @@ impl<HttpMiddleware, RpcMiddleware> Builder<HttpMiddleware, RpcMiddleware> {
-	/// Whether batch requests are supported by this server or not.
-	pub(crate) batch_requests_config: BatchRequestConfig,
-	/// Subscription ID provider.
-	pub(crate) id_provider: Arc<dyn IdProvider>,
-	/// Ping configuration.
-	pub(crate) ping_config: PingConfig,
+	methods: Methods,
 	/// Stop handle.
-	pub(crate) stop_handle: StopHandle,
+	stop_handle: StopHandle,
 	/// Connection ID
-	pub(crate) conn_id: u32,
-	/// Handle to hold a `connection permit`.
-	pub(crate) conn: Arc<OwnedSemaphorePermit>,
-	/// Enable HTTP.
-	pub(crate) enable_http: bool,
-	/// Enable WS.
-	pub(crate) enable_ws: bool,
-	/// Number of messages that server is allowed `buffer` until backpressure kicks in.
-	pub(crate) message_buffer_capacity: u32,
+	conn_id: u32,
+	/// Connection guard.
+	conn_guard: ConnectionGuard,
+	/// ServerConfig
+	server_cfg: ServerConfig,
 }
 
-/// JsonRPSee service compatible with `tower`.
+/// jsonrpsee tower service
+///
+/// This will enable both `http_middleware` and `rpc_middleware`
+/// if you want only to enable `rpc_middleware` then [`TowerServiceNoHttp`]
+/// can be used.
+#[derive(Debug)]
+pub struct TowerService<RpcMiddleware, HttpMiddleware> {
+	rpc_middleware: TowerServiceNoHttp<RpcMiddleware>,
+	http_middleware: tower::ServiceBuilder<HttpMiddleware>,
+}
+
+impl<RpcMiddleware, HttpMiddleware> hyper::service::Service<hyper::Request<hyper::Body>>
+	for TowerService<RpcMiddleware, HttpMiddleware>
+where
+	RpcMiddleware: for<'a> tower::Layer<RpcService> + Clone,
+	<RpcMiddleware as Layer<RpcService>>::Service: Send + Sync + 'static,
+	for<'a> <RpcMiddleware as Layer<RpcService>>::Service: RpcServiceT<'a>,
+	HttpMiddleware: Layer<TowerServiceNoHttp<RpcMiddleware>> + Send + 'static,
+	<HttpMiddleware as Layer<TowerServiceNoHttp<RpcMiddleware>>>::Service: Send
+		+ Service<
+			hyper::Request<hyper::Body>,
+			Response = hyper::Response<hyper::Body>,
+			Error = Box<(dyn StdError + Send + Sync + 'static)>,
+		>,
+	<<HttpMiddleware as Layer<TowerServiceNoHttp<RpcMiddleware>>>::Service as Service<hyper::Request<hyper::Body>>>::Future:
+		Send + 'static,
+{
+	type Response = hyper::Response<hyper::Body>;
+	type Error = Box<dyn StdError + Send + Sync + 'static>;
+	type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+	/// Opens door for back pressure implementation.
+	fn poll_ready(&mut self, _: &mut std::task::Context) -> Poll<Result<(), Self::Error>> {
+		Poll::Ready(Ok(()))
+	}
+
+	fn call(&mut self, request: hyper::Request<hyper::Body>) -> Self::Future {
+		Box::pin(self.http_middleware.service(self.rpc_middleware.clone()).call(request))
+	}
+}
+
+/// jsonrpsee tower service.
+///
+/// This will not apply specific HTTP middleware that
+/// can be enabled by [`Builder::set_http_middleware`](method@Builder::set_http_middleware).
 ///
 /// # Note
 /// This is similar to [`hyper::service::service_fn`].
 #[derive(Debug, Clone)]
-pub struct TowerService<L> {
+pub struct TowerServiceNoHttp<L> {
```

Comment:
bikeshed on naming,

I think this shouldn't be used by external users because the tower middleware from ServerBuilder isn't used then

Reply:
changed this to jsonrpsee internal thingy now